### PR TITLE
CumulativeAverage refactor

### PR DESCRIPTION
**Pull request by myron**
This refactors CumulativeAverage() by removing subclassing from Cumulative() to have much less overhead, and adds additionan functionality to specify count. 
(as discussed here https://github.com/Project-MONAI/MONAI/pull/5380)

CumulativeAverage previously was not used almost anywhere, except for MIL tutorial. Name of class methods .aggregate() and .append() are preserved, so it should work the same way as before (non-breaking).

New tests are added (especially to test in multi-gpu DDP)

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by myron** [comment]
seems it passed all the tests  @Nic-Ma

**Comment by myron** [comment]
@wyli @Nic-Ma  any objections to merging it? (this class was originally created due to my request, and this PR improves it)

File: monai/metrics/cumulative_average.py [hunk 1]
```diff
@@ -9,64 +9,162 @@
...
+        """
+
+        self.is_distributed = dist.is_available() and dist.is_initialized()
+        self.group = None
+        self.device = None
+
+        if self.is_distributed:
+            dist_cuda = dist.get_backend() == dist.Backend.NCCL
+            if ensure_cpu_group and dist_cuda:
+                self.group = dist.new_group(backend="gloo")  # create a new cpu group
```
Comment (wyli) [inline monai/metrics/cumulative_average.py:57]: no sure if this class should manage the process group, this should be decoupled. also `new_group` takes additional parameters, it's not easy to make a suitable assumption about the system environment and have a robust implementation within this class

Comment (myron) [reply monai/metrics/cumulative_average.py]: @wyli good point, thank you. I've updated the code,  to have the process group as an option (if a user wants to provide a custom group).

**Comment by myron** [comment]
@ericspod  I've updated it according to your comments in https://github.com/Project-MONAI/MONAI/pull/5380
Can you please check

File: monai/metrics/cumulative_average.py [hunk 2]
```diff
@@ -9,64 +9,160 @@
...
+        self.is_distributed = dist.is_available() and dist.is_initialized()
+        self.device = None
+
+        if self.is_distributed:
+            self.group = group
+            if dist.get_backend(group=group) == dist.Backend.NCCL:
+                self.device = torch.device("cuda", dist.get_rank(group=group))
+        else:
+            self.group = None
+
```
Comment (ericspod) [inline monai/metrics/cumulative_average.py:58]:
```suggestion
if self.device is None:
    self.device=torch.device("cpu")
```

File: monai/metrics/cumulative_average.py [hunk 3]
```diff
@@ -9,64 +9,160 @@
...
         """
-        Reset all the running status, including buffers, sum, not nans count, etc.
+        Ensure torch.Tensor float format and optionally copy to the proper gpu device
+        """
+        x: torch.Tensor = a.detach() if isinstance(a, torch.Tensor) else torch.tensor(a)
+
+        if self.device is not None:
+            x = x.to(device=self.device, dtype=torch.float)
+        else:
+            x = x.cpu().float()
```
Comment (ericspod) [inline monai/metrics/cumulative_average.py:79]: If `self.device` isn't ever `None`:
```suggestion
        x = x.to(device=self.device, dtype=torch.float)
```

File: monai/metrics/cumulative_average.py [hunk 4]
```diff
@@ -9,64 +9,170 @@
...
 
-        return self.sum / self.not_nans
+        val = self.proper_tensor(val)
+        count = self.proper_tensor(count)
+
+        val_count = val * count
+        nfin = ~torch.isfinite(val_count)
+        if torch.any(nfin):
+            # non-finite numbers may indicate some errors in the user code
+            print("non-finite numbers received", val, count)
```
Comment (ericspod) [inline monai/metrics/cumulative_average.py:170]: We shouldn't have print statements in regular code since at times we would want programs to be silent. Instead if it's important to state this we should write to the log or issue a warning with `warnings.warn`.

File: monai/metrics/cumulative_average.py (hunk 2 above)
Comment (myron) [reply monai/metrics/cumulative_average.py:59]: updated

File: monai/metrics/cumulative_average.py (hunk 3 above)
Comment (myron) [reply monai/metrics/cumulative_average.py]: updated

File: monai/metrics/cumulative_average.py (hunk 4 above)
Comment (myron) [reply monai/metrics/cumulative_average.py]: okay, I 've changed it to warnings. thank you

**Comment by myron** [comment]
@ericspod thank you for  the quick review. I've updated the code according to your comments. plz check again.

File: monai/metrics/cumulative_average.py [hunk 5]
```diff
@@ -9,64 +9,168 @@
...
+    def reset(self) -> None:
         """
-        Reset all the running status, including buffers, sum, not nans count, etc.
+        Reset all the running status
 
         """
-        super().reset()
-        self.sum = None
-        self.not_nans = None
+        self.val: torch.Tensor = 0  # type: ignore
```
Comment (ericspod) [inline monai/metrics/cumulative_average.py:70]: It's slightly inefficient but using actual tensors here instead of 0 would be more consistent and not need the type ignore comment, but it's not critical.

**Review by ericspod — APPROVED** [review]

**Comment by ericspod** [comment]
I'm good with this if others are?

File: monai/metrics/cumulative_average.py (hunk 5 above)
Comment (myron) [reply monai/metrics/cumulative_average.py:70]: yes, I tried it, using actual torch tensor 0, but we don't know the shape of val, sum, count before stating to accumulate, and later self.sum += some_tensor fails, since in place increment not possible with different tensor shapes (even if incrementing 0 dim tensor).  So we'd have to use self.sum = self.sum + some_tensor, out of place increment, which is slightly bigger overhead. It's seems simpler to have it as is.

File: monai/metrics/cumulative_average.py [hunk 6]
```diff
@@ -9,64 +9,168 @@
...
+        self.sum: torch.Tensor = 0  # type: ignore
+        self.count: torch.Tensor = 0  # type: ignore
 
-    def aggregate(self):
+    def proper_tensor(self, a: Any) -> torch.Tensor:
         """
-        Sync data from all the ranks and compute the average value with previous sum value.
+        Ensure torch.Tensor float format and optionally copy to the proper gpu device
+        """
+        x: torch.Tensor = a.detach() if isinstance(a, torch.Tensor) else torch.tensor(a)
```
Comment (wyli) [inline monai/metrics/cumulative_average.py:78]: Normally the metrics module shouldn't change the device of the input if we follow pytorch's api style. The user should prepare the input beforehand accordingly in a compatible shape/type/device... If this is 100% needed, we may have to change all the relevant APIs just to me consistent

Comment (myron) [reply monai/metrics/cumulative_average.py:78]: this line is just to ensure the the avergage value is on the correct device ( CPU vs GPU).  If a user provide it as a simple constant or a list (on cpu and we're in DDP on gpu), it'll be converted to proper GPU device. And otherwise if we're not in DDP, it will ensure we're maintaining average on CPU.  

many pytorch api accepts input as either a constant or a tensor, and support shape broadcasting. In either case this  self.proper_tensor()  is just an internal sub-function.

Comment (wyli) [reply monai/metrics/cumulative_average.py:78]: sorry, the comment is meant to be on the the `self.device` usage in general in this class

Comment (myron) [reply monai/metrics/cumulative_average.py:78]: @wyli thank you for your comment. the self.device is set according to the processing group. If the user specified group on CPU, then device is on CPU, if the user specified it on GPU (or it's default group on GPU/ NCCL) then the device is GPU.  We need to insure the proper device to be able to run multiprocessing reduction on proper Processing group in DDP.  Please let me know if there is something specific you'd like me to change in the code.  thank you

Comment (myron) [reply monai/metrics/cumulative_average.py:78]: @wyli , should I mark this conversation as resolved? thank you

Comment (wyli) [reply monai/metrics/cumulative_average.py:78]: Thanks, I don't think this class should assign the devices, it doesn't even consider the difference between rank vs local rank.

Comment (myron) [reply monai/metrics/cumulative_average.py]: @wyli alright, I've removed self.device , it will be up to a user now to use the correct device.  Let me know if you have any other objections.
